server/backend/api/app/crud/services/ext_app_user_search_service.py

```diff
@@ -0,0 +1,218 @@
+import logging
+from http import HTTPStatus
+from typing import List
+
+from api.app.constants import (ERROR_CODE_INVALID_OPERATION, EXT_MIN_PAGE,
+                               EXT_MIN_PAGE_SIZE, IDPType, ScopeType, UserType)
+from api.app.models.model import FamRole, FamUser, FamUserRoleXref
+from api.app.schemas.ext.pagination import (ExtUserSearchPagedResultsSchema,
+                                            ExtUserSearchParamSchema)
+from api.app.schemas.ext.user_search import (ExtApplicationUserSearchGetSchema,
+                                             ExtApplicationUserSearchSchema,
+                                             ExtRoleWithScopeSchema)
+from api.app.schemas.requester import RequesterSchema
+from api.app.utils.utils import raise_http_exception
+from sqlalchemy import and_, func, or_, select
+from sqlalchemy.orm import Session, joinedload
+from sqlalchemy.sql.selectable import Select
+
+from .ext_api_interface import ExtAPIInterface
+
+LOGGER = logging.getLogger(__name__)
+
+class ExtAppUserSearchService(ExtAPIInterface):
+    """
+    Service to handle external application user search requests for external API calls.
+    The service only allows requesters with proper application role (with call_api permission)
+    to perform the search.
+    Ref API spec at https://apps.nrs.gov.bc.ca/int/confluence/display/FSAST1/Users+Search+API
+    """
+
+    def __init__(self, db: Session, requester: RequesterSchema, application_id: int, *args, **kwargs):
+        super().__init__(requester, application_id, db=db, *args, **kwargs)
+
+    def search_users(
+        self,
+        page_params: ExtUserSearchParamSchema,
+        filter_params: ExtApplicationUserSearchSchema
+    ) -> ExtUserSearchPagedResultsSchema:
+        """
+        For external Consumer API only.
+        Searches users associated with the application based on filter_params and paginates results based on page_params.
+        """
+        if not self.is_request_allowed():
+            error_msg = ("Programming error. This method is for external API only. "
+                         "Router should have already checked the requester call API permission.")
+            raise_http_exception(
+                status_code=HTTPStatus.INTERNAL_SERVER_ERROR,
+                error_code=ERROR_CODE_INVALID_OPERATION,
+                error_msg=error_msg,
+            )
+
+        LOGGER.debug(f"Searching users with filter_params: {filter_params}, page_params: {page_params}")
+
+        # Base select statement
+        user_role_stmt: Select = (
+            select(FamUser)
+            .join(FamUserRoleXref, FamUser.user_id == FamUserRoleXref.user_id)
+            .join(FamRole, FamUserRoleXref.role_id == FamRole.role_id)
+            .where(FamRole.application_id == self.application_id)
```

Comment:
Just to confirm, all these queries for roles are limiting it to just roles for the application matching the client of the user token supplied to this endpoint?